From: TomLiu <tomml.liu@mediatek.com>
Date: Fri, 12 Apr 2024 17:07:00 +0000
Subject: [PATCH] wifi: mt76: mt7915: get airtime from MCU event

Retrieve per-STA airtime via a single MCU event instead of polling
individual WTBL registers per station.

Signed-off-by: TomLiu <tomml.liu@mediatek.com>
Signed-off-by: Chad Monroe <chad@monroe.io>
---
 mt76_connac_mcu.h |   12 ++++++++
 mt7915/mac.c      |   66 +++-----------------------------------------
 mt7915/mcu.c      |   58 ++++++++++++++++++++++++++++++++++++++
 mt7915/mcu.h      |   16 ++++++++++
 mt7915/mt7915.h   |    1
 5 files changed, 92 insertions(+), 61 deletions(-)

--- a/mt76_connac_mcu.h
+++ b/mt76_connac_mcu.h
@@ -1081,6 +1081,7 @@ enum {
 	MCU_EXT_EVENT_BCC_NOTIFY = 0x75,
 	MCU_EXT_EVENT_WF_RF_PIN_CTRL = 0x9a,
 	MCU_EXT_EVENT_MURU_CTRL = 0x9f,
+	MCU_EXT_EVENT_GET_ALL_STA_STATS = 0xb5,
 };
 
 /* unified event table */
@@ -1321,6 +1322,7 @@ enum {
 	MCU_EXT_CMD_GROUP_PRE_CAL_INFO = 0xab,
 	MCU_EXT_CMD_DPD_PRE_CAL_INFO = 0xac,
 	MCU_EXT_CMD_PHY_STAT_INFO = 0xad,
+	MCU_EXT_CMD_GET_ALL_STA_STATS = 0xb5,
 	MCU_EXT_CMD_EDCCA = 0xba,
 	MCU_EXT_CMD_WF_RF_PIN_CTRL = 0xbd,
 	MCU_EXT_CMD_SWLNA_ACI_CTRL = 0xc0,
@@ -1328,6 +1330,16 @@ enum {
 	MCU_EXT_CMD_LPI_CTRL = 0xc8,
 };
 
+enum MCU_EXT_CMD_ALL_STA_INFO_TAG {
+	MCU_PHY_PER_STA_TX_RATE = 0x1,
+	MCU_PHY_PER_STA_TX_STAT = 0x2,
+	MCU_PHY_RX_STAT = 0x03,
+	MCU_PHY_PER_STA_TXRX_ADM_STAT = 0x04,
+	MCU_PHY_PER_STA_TXRX_AIR_TIME = 0x05,
+	MCU_PHY_PER_STA_DATA_TX_RETRY_COUNT = 0x06,
+	MCU_PHY_PER_STA_GI_MODE = 0x07,
+};
+
 enum {
 	MCU_UNI_CMD_DEV_INFO_UPDATE = 0x01,
 	MCU_UNI_CMD_BSS_INFO_UPDATE = 0x02,
--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -89,18 +89,9 @@ u32 mt7915_mac_wtbl_lmac_addr(struct mt7
 
 static void mt7915_mac_sta_poll(struct mt7915_dev *dev)
 {
-	static const u8 ac_to_tid[] = {
-		[IEEE80211_AC_BE] = 0,
-		[IEEE80211_AC_BK] = 1,
-		[IEEE80211_AC_VI] = 4,
-		[IEEE80211_AC_VO] = 6
-	};
-	struct ieee80211_sta *sta;
 	struct mt7915_sta *msta;
 	struct rate_info *rate;
-	u32 tx_time[IEEE80211_NUM_ACS], rx_time[IEEE80211_NUM_ACS];
 	LIST_HEAD(sta_poll_list);
-	int i;
 
 	rcu_read_lock();
 
@@ -109,7 +100,6 @@ static void mt7915_mac_sta_poll(struct m
 	spin_unlock_bh(&dev->mt76.sta_poll_lock);
 
 	while (true) {
-		bool clear = false;
 		u32 addr, val;
 		u16 idx;
 		s8 rssi[4];
@@ -126,57 +116,6 @@ static void mt7915_mac_sta_poll(struct m
 		spin_unlock_bh(&dev->mt76.sta_poll_lock);
 
 		idx = msta->wcid.idx;
-
-		/* refresh peer's airtime reporting */
-		addr = mt7915_mac_wtbl_lmac_addr(dev, idx, 20);
-
-		for (i = 0; i < IEEE80211_NUM_ACS; i++) {
-			u32 tx_last = msta->airtime_ac[i];
-			u32 rx_last = msta->airtime_ac[i + 4];
-
-			msta->airtime_ac[i] = mt76_rr(dev, addr);
-			msta->airtime_ac[i + 4] = mt76_rr(dev, addr + 4);
-
-			if (msta->airtime_ac[i] <= tx_last)
-				tx_time[i] = 0;
-			else
-				tx_time[i] = msta->airtime_ac[i] - tx_last;
-
-			if (msta->airtime_ac[i + 4] <= rx_last)
-				rx_time[i] = 0;
-			else
-				rx_time[i] = msta->airtime_ac[i + 4] - rx_last;
-
-			if ((tx_last | rx_last) & BIT(30))
-				clear = true;
-
-			addr += 8;
-		}
-
-		if (clear) {
-			mt7915_mac_wtbl_update(dev, idx,
-					       MT_WTBL_UPDATE_ADM_COUNT_CLEAR);
-			memset(msta->airtime_ac, 0, sizeof(msta->airtime_ac));
-		}
-
-		if (!msta->wcid.sta)
-			continue;
-
-		sta = container_of((void *)msta, struct ieee80211_sta,
-				   drv_priv);
-		for (i = 0; i < IEEE80211_NUM_ACS; i++) {
-			u8 queue = mt76_connac_lmac_mapping(i);
-			u32 tx_cur = tx_time[queue];
-			u32 rx_cur = rx_time[queue];
-			u8 tid = ac_to_tid[i];
-
-			if (!tx_cur && !rx_cur)
-				continue;
-
-			ieee80211_sta_register_airtime(sta, tid, tx_cur,
-						       rx_cur);
-		}
-
 		rate = &msta->wcid.rate;
 
 		switch (rate->bw) {
@@ -2018,11 +1957,14 @@ void mt7915_mac_work(struct work_struct
 {
 	struct mt7915_phy *phy;
 	struct mt76_phy *mphy;
+	struct mt76_dev *mdev;
 
 	mphy = (struct mt76_phy *)container_of(work, struct mt76_phy,
 					       mac_work.work);
 	phy = mphy->priv;
 
+	mdev = mphy->dev;
+
 	mutex_lock(&mphy->dev->mutex);
 
 	mt76_update_survey(mphy);
@@ -2034,6 +1976,8 @@ void mt7915_mac_work(struct work_struct
 
 		if (phy->dev->muru_debug)
 			mt7915_mcu_muru_debug_get(phy);
+
+		mt7915_mcu_get_all_sta_info(mdev, MCU_PHY_PER_STA_TXRX_AIR_TIME);
 	}
 
 	if (++phy->stats_work_count == 10) {
--- a/mt7915/mcu.c
+++ b/mt7915/mcu.c
@@ -469,6 +469,57 @@ mt7915_mcu_rx_ps_sync(struct mt7915_dev
 }
 
 static void
+mt7915_mcu_rx_all_sta_info_event(struct mt7915_dev *dev, struct sk_buff *skb)
+{
+	struct mt7915_mcu_all_sta_info_event *res;
+	u16 sta_num;
+	u16 i;
+
+	res = (struct mt7915_mcu_all_sta_info_event *)skb->data;
+	if (skb->len < sizeof(*res))
+		return;
+
+	sta_num = le16_to_cpu(res->sta_num);
+	if (sta_num > (skb->len - sizeof(*res)) / sizeof(res->airtime[0]))
+		sta_num = (skb->len - sizeof(*res)) / sizeof(res->airtime[0]);
+
+	rcu_read_lock();
+	for (i = 0; i < sta_num; i++) {
+		u8 ac;
+		u16 wlan_idx;
+		struct mt76_wcid *wcid;
+		struct ieee80211_sta *sta;
+		u32 tx_airtime, rx_airtime;
+
+		switch (res->tag) {
+		case MCU_PHY_PER_STA_TXRX_AIR_TIME:
+			wlan_idx = le16_to_cpu(res->airtime[i].wlan_idx);
+			wcid = mt76_wcid_ptr(dev, wlan_idx);
+
+			sta = wcid_to_sta(wcid);
+			if (!sta)
+				continue;
+
+			for (ac = IEEE80211_AC_VO; ac < IEEE80211_NUM_ACS; ++ac) {
+				u8 lmac_ac = mt76_connac_lmac_mapping(ac);
+
+				tx_airtime = le32_to_cpu(res->airtime[i].tx[lmac_ac]);
+				rx_airtime = le32_to_cpu(res->airtime[i].rx[lmac_ac]);
+
+				wcid->stats.tx_airtime += tx_airtime;
+				wcid->stats.rx_airtime += rx_airtime;
+				ieee80211_sta_register_airtime(sta, mt76_ac_to_tid(ac),
+							       tx_airtime, rx_airtime);
+			}
+			break;
+		default:
+			break;
+		}
+	}
+	rcu_read_unlock();
+}
+
+static void
 mt7915_mcu_rx_ext_event(struct mt7915_dev *dev, struct sk_buff *skb)
 {
 	struct mt76_connac2_mcu_rxd *rxd;
@@ -493,6 +544,9 @@ mt7915_mcu_rx_ext_event(struct mt7915_de
 	case MCU_EXT_EVENT_PS_SYNC:
 		mt7915_mcu_rx_ps_sync(dev, skb);
 		break;
+	case MCU_EXT_EVENT_GET_ALL_STA_STATS:
+		mt7915_mcu_rx_all_sta_info_event(dev, skb);
+		break;
 	default:
 		break;
 	}
@@ -524,6 +578,7 @@ void mt7915_mcu_rx_event(struct mt7915_d
 	     rxd->ext_eid == MCU_EXT_EVENT_ASSERT_DUMP ||
 	     rxd->ext_eid == MCU_EXT_EVENT_PS_SYNC ||
 	     rxd->ext_eid == MCU_EXT_EVENT_BCC_NOTIFY ||
+	     rxd->ext_eid == MCU_EXT_EVENT_GET_ALL_STA_STATS ||
 	     !rxd->seq) &&
 	     !(rxd->eid == MCU_CMD_EXT_CID &&
 	       rxd->ext_eid == MCU_EXT_EVENT_WA_TX_STAT))
@@ -5020,3 +5075,16 @@ int mt7915_mcu_thermal_debug(struct mt79
 	return mt76_mcu_send_msg(&dev->mt76, MCU_EXT_CMD(THERMAL_DEBUG), &req,
 				 sizeof(req), true);
 }
+
+int mt7915_mcu_get_all_sta_info(struct mt76_dev *dev, u16 tag)
+{
+	struct {
+		u8 tag;
+		u8 rsv[3];
+	} __packed req = {
+		.tag = tag,
+	};
+
+	return mt76_mcu_send_msg(dev, MCU_EXT_CMD(GET_ALL_STA_STATS),
+				&req, sizeof(req), false);
+}
--- a/mt7915/mcu.h
+++ b/mt7915/mcu.h
@@ -583,6 +583,22 @@ enum {
 	SER_RECOVER
 };
 
+struct mt7915_mcu_all_sta_info_event {
+	struct mt76_connac2_mcu_rxd rxd;
+
+	u8 tag;
+	u8 more;
+	__le16 sta_num;
+
+	union {
+		struct {
+			__le16 wlan_idx;
+			__le32 tx[IEEE80211_NUM_ACS];
+			__le32 rx[IEEE80211_NUM_ACS];
+		} airtime[0] __packed;
+	} __packed;
+} __packed;
+
 #define MT7915_MAX_BEACON_SIZE		1308
 #define MT7915_BEACON_UPDATE_SIZE	(sizeof(struct sta_req_hdr) +	\
 					 sizeof(struct bss_info_bcn) +	\
--- a/mt7915/mt7915.h
+++ b/mt7915/mt7915.h
@@ -613,6 +613,7 @@ int mt7915_mcu_set_thermal_protect(struc
 int mt7915_mcu_get_tx_rate(struct mt7915_phy *phy, u16 wcidx);
 int mt7915_mcu_get_rx_rate(struct mt7915_phy *phy, struct ieee80211_vif *vif,
 			   struct ieee80211_sta *sta, struct rate_info *rate);
+int mt7915_mcu_get_all_sta_info(struct mt76_dev *dev, u16 tag);
 int mt7915_mcu_rdd_background_enable(struct mt7915_phy *phy,
 				     struct cfg80211_chan_def *chandef);
 int mt7915_mcu_wed_wa_tx_stats(struct mt7915_dev *dev, u16 wcid);
